timemachine/fe/single_topology.py

```diff
@@ -0,0 +1,953 @@
+            assert len(atoms) == 2
+            k0, k1 = atoms
+            restraint_angle_idxs.append((i, j, k0))
+            restraint_angle_params.append((100.0, (2.0 / 3.0) * np.pi))
+            restraint_angle_idxs.append((i, j, k1))
+            restraint_angle_params.append((100.0, (2.0 / 3.0) * np.pi))
+        elif anchor_full_geometry == LocalGeometry.G4_TETRAHEDRAL:
+            # diagram:
+            #     k0
+            #    .
+            # i-j . k2
+            #    .
+            #     k1
+            # add three angles: (i,j,k0) and (i,j,k1) and (i,j,k2)
+            assert anchor in chiral_atoms
+            atoms = find_attached_dummy_atoms(dummy_group, mol_b_bond_idxs, anchor)
+            assert len(atoms) == 3
+            k0, k1, k2 = atoms
+            restraint_angle_idxs.append((i, j, k0))
+            restraint_angle_params.append((100.0, 1.91))
+            restraint_angle_idxs.append((i, j, k1))
+            restraint_angle_params.append((100.0, 1.91))
+            restraint_angle_idxs.append((i, j, k2))
+            restraint_angle_params.append((100.0, 1.91))
+        else:
+            assert 0, "Illegal Geometry"
+    elif anchor_core_geometry == LocalGeometry.G2_KINK:
+        # type a
+        #       \
+        #        j
+        #       /
+        #      b
+        # a and b are core atoms, j being the anchor
+
+        j = anchor
+
+        if anchor_full_geometry == LocalGeometry.G3_PYRAMIDAL:
+            # type
+            #    a - j
+            #       / .
+            #      b   k
+            # find stable core atoms next to anchor that we can build angle restraints off of.
+            a, b = nbs_1
+
+            assert check_bond_angle_stability(
+                a, j, b, mol_b_bond_idxs, mol_b_bond_params, mol_b_angle_idxs, mol_b_angle_params
+            )
+            assert check_bond_angle_stability(
+                core_b_to_a[a],
+                core_b_to_a[j],
+                core_b_to_a[b],
+                mol_a_bond_idxs,
+                mol_a_bond_params,
+                mol_a_angle_idxs,
+                mol_a_angle_params,
+            )
+
+            # a and b are core atoms, j being the anchor
+            atoms = find_attached_dummy_atoms(dummy_group, mol_b_bond_idxs, anchor)
+            assert len(atoms) == 1
+            k = atoms[0]
+
+            if anchor in chiral_atoms:
+                assert 0, "Junction atom is chiral and G3_PYRAMIDAL."
+                restraint_cross_angle_idxs.append(((j, a), (j, b), (j, k)))
+                restraint_cross_angle_params.append(1000.0)
+            else:
+                # planarize so we can enhance sample both chiral states using a centroid
+                # type a
+                #       \
+                #      c.j.k <- angle (c,j,k) = 0.0
+                #       /
+                #      b
+                restraint_centroid_angle_idxs.append((tuple(sorted((a, b))), j, k))
```

Comment:
could we use `canonicalize_bond` here?

Reply:
`(a,b)` aren't actually bonded to each other. Rather they're bonded to the junction atom `j`. The `sorted()` is purely a convenience for testing purposes. So I didn't want to confuse the reader into thinking they were actually bonds.